### DISCUSSION TITLE
🚀 Add sub-directory awareness to File addon

### DISCUSSION THREAD

**Discussion by jcogs-design** [ExpressionEngine/ExpressionEngine#3820] [Feature Requests]
### Suggest an idea for ExpressionEngine

The [File addon](https://docs.expressionengine.com/latest/add-ons/file.html) is useful, but seems not to be aware of / able to work with EE7 subdirectories.
Interrogating for `{directory_id}` returns the parent directory id, but no information about the sub-folder that might be in use.
Similarly it appears impossible to get the add-on to list the content of a sub-folder.
Without this ability to interrogate images for location is limiting. There appears to be no other way to determine the location of an image within the file structure, and no other way to list out the content of a sub-folder.

It would seem helpful if awareness of EE7 file system new features was added to this module.

**Comment by intoeetive** [comment]
@jcogs-design there is `{folder_id}` variable that you can use (and that we need to document)

Also I agree that ability to limit by folder would be useful here

**Comment by jcogs-design** [comment]
It doesn't look like the `{folder_id}` variable is being produced when interrogating a Files field with a single file selected. Should it be? If not, how would I work out what `{folder_id}` applies to an image selected in a Files field?

**Comment by intoeetive** [comment]
I think I am confused. Are you talking about file module (i.e. `{exp:file:entries}` tag), or the File fieldtype?

`{folder_id}` does not seem to be available as fieldtype variable. Perhaps worth adding it (but out of pure interest, do you have a practical example / template chunk for using it?)

**Comment by jcogs-design** [comment]
It also looks like the `{folder_id}` variable in the File add-on returns the wrong number ... 😢 
What I have done to work-around these limitations is as follows:

- Content editor chooses an image taken from the sub-folder and saves in a File field in CP
- Export the `{file_id}` of this image from File variable 
- Use the `{file_id}` to load up image in the File add-on
- Capture the values for `{directory_id}` and `{folder_id}` from that listing
- List the content of the directory using the File add-on and run a conditional to only accept images that have `{folder_id}` that matches the one obtained from the test image

However, this code:
```
    {exp:file:entries file_id="{embed:file_id}"}
    {layout:set name="directory"{directory_id}{/layout:set}
    {layout:set name="folder"}{folder_id}{/layout:set}
    {/exp:file:entries}
```
generates this layout variable output
```
array (
  'directory' => '20',
  'folder' => '20',
)
```
by comparison when viewing the relevant sub-folder in CP the URL has the following content:
```
/cp/files/directory/20&directory_id=434
```
For other sub-folders in the same file directory the `directory_id` value changes - so I am guessing that this is (despite variation compared to File add-on in naming) is the `{folder_id}` value.

**Comment by intoeetive** [comment]
@jcogs-design please check this out

code: https://github.com/ExpressionEngine/ExpressionEngine/pull/3823

docs: https://github.com/ExpressionEngine/ExpressionEngine-User-Guide/pull/751

If you find it working well, please approve using "review changes" button on Files tab for the PRs. thank you!
